### PR TITLE
nn <> cudnn conversion

### PR DESCRIPTION
**Pull request by szagoruyko**
Easy conversion function:
```lua
net = cudnn.convert(net, cudnn)
```
will go recursively over a network and substitute all compatible layers to a given backend. The parameters are not cloned, only references are copied.

Going to add some tests, tested locally so far.

### CODE REVIEW

**Comment by Kaixhin** [comment]
Very nice. More conversions to come in [R4](https://github.com/soumith/cudnn.torch/tree/R4), which would make this even more convenient!

**Comment by szagoruyko** [comment]
added conversion tests to all modules:

| Layer | `nn` impl | conversion test |
| ----- |:-----:|:-----:|
| LogSoftMax_batch | x | x |
| LogSoftMax_single | x | x |
| ReLU_batch | x | x |
| ReLU_single | x | x |
| Sigmoid_batch | x | x |
| Sigmoid_single | x | x |
| SoftMax_batch | x | x |
| SoftMax_single | x | x |
| SpatialAveragePooling_batch | x | x |
| SpatialAveragePooling_single | x | x |
| SpatialConvolution_backward_batch | x | x |
| SpatialConvolution_backward_single | x | x |
| SpatialConvolution_forward_batch | x | x |
| SpatialConvolution_forward_single | x | x |
| SpatialCrossEntropyCriterion | | |
| SpatialCrossMapLRN_batch | x | x |
| SpatialLogSoftMax | | |
| SpatialMaxPooling_batch | x | x |
| SpatialMaxPooling_single | x | x |
| Tanh_batch | x | x |
| Tanh_single | x | x |
| VolumetricConvolution_backward_single | x | x |
| VolumetricConvolution_forward_single | x | x |
| VolumetricMaxPooling_batch | x | x |
| VolumetricMaxPooling_single | x | x |

@soumith please review https://github.com/soumith/cudnn.torch/pull/83 and this one and we can merge

**Comment by szagoruyko** [comment]
This should clean all the backend switch code like https://github.com/szagoruyko/cifar.torch/blob/master/models/vgg_bn_drop.lua#L6 or 
https://github.com/soumith/imagenet-multiGPU.torch/blob/master/models/alexnet_cudnn.lua
so the workflow is simple:
1. define a model in `nn`
2. if cudnn can be included, call `cudnn.convert`
3 (optional). call `cudnn.convert(net, nn)` before `torch.save`

There are still some differences in `SpatialConvolution` with groups implementation, I think we should add a dummy groups option to `nn` and assert in `updateOutput` that groups are not implemented.
Also, `SpatialLogSoftMax` and `SpatialCrossEntropyCriterion` don't have their `nn` equivalents.

**Comment by fmassa** [comment]
I'll send a PR with a SpatialLogSoftMax, so there will be full conversion support between cudnn and nn (SpatialCrossEntropyCriterion can be copied as is afterwards).
I'll also send a PR with groups option for nn, it will come together with the THNN refactoring.
I'll come back from holidays next week and I'll prepare the PRs

**Comment by szagoruyko** [comment]
@fmassa thanks Francisco

I think we can merge this and add the features Francisco mentioned later.

File: convert.lua [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  return replace(net, function(x)
+    local y = 0
+    local src = dst == nn and cudnn or nn
+    local src_prefix = src == nn and 'nn.' or 'cudnn.'
+    local dst_prefix = dst == nn and 'nn.' or 'cudnn.'
+
+    local function convert(v)
+      local y = {}
+      torch.setmetatable(y, dst_prefix..v)
+      if v == 'ReLU' then y = dst.ReLU() end -- because parameters
```
Comment (fmassa) [inline convert.lua]: I still don't get why we need this line. Could you explain a bit ?

Comment (szagoruyko) [inline convert.lua]: @fmassa `nn.ReLU` is implemented via `nn.Threshold` https://github.com/torch/nn/blob/master/Threshold.lua and has additional `threshold` and `val` fields absent in `cudnn.ReLU`

Comment (fmassa) [inline convert.lua]: I see. Yeah, it's better to add this case here than adding to `ReLU` forward an initialization of `threshold` and `val` if not present. It's clearer this way. Thanks !

**Comment by fmassa** [comment]
Did you tried saving a `nn` model which was converted from `cudnn` ? Don't you have messages like "the object can't be serialized" because we are not freeing the buffers like in https://github.com/soumith/cudnn.torch/blob/master/SpatialConvolution.lua#L447-L458 ?

**Comment by szagoruyko** [comment]
yes, there warnings are there. let me add a cleaner

**Comment by szagoruyko** [comment]
@fmassa added `cleanDesc` function and `scaleT` too

**Comment by szagoruyko** [comment]
everyone ok with merging this?

**Comment by fmassa** [comment]
Could add some docs ? Except from that, I think it's good.

**Comment by szagoruyko** [comment]
added a note in readme and squashed

File: README.md [hunk 2]
```diff
@@ -59,6 +60,27 @@ cudnn.verbose = true -- this prints out some more verbose information useful for
 ```
 by default, `cudnn.verbose` is set to `false`.
 
+### Conversion between `cudnn` and `nn`
+
+Conversion is done by `cudnn.convert` function which takes a network and backend arguments and goes over
```
Comment (fmassa) [inline README.md]: Could you also mention (or add an example) that you can convert to both `nn` and `cudnn`  backends ? It's not clear from the README that it's possible to go from `cudnn` to `nn`.

Comment (szagoruyko) [inline README.md]: updated, thanks

**Comment by fmassa** [comment]
Looks good to me. Thanks Sergey !
@soumith can we merge it ?

**Pull request merged**